src/v/datalake/translation/scheduling.h

```diff
@@ -0,0 +1,357 @@
+     * \ref scheduler_notifications::notify_ready notification.
+     *
+     * Once finished, the translator should release all the
+     * resources and notify the scheduler via
+     * \ref scheduling_notifications::notify_done.
+     */
+    virtual void start_translation(clock::duration time_slice) = 0;
+
+    /**
+     * Invoked by the scheduler to preempt an inprogress translation. All
+     * resources should be released and the translator should invoke \ref
+     * scheduling_notifications::notify_done when done.
+     */
+    virtual void stop_translation() = 0;
+};
+
+std::ostream& operator<<(std::ostream&, const translator&);
+
+struct executor;
+class scheduler;
+
+/**
+ * Encapsulates all the scheduling state for a single translator.
+ *   ┌───────┐
+ *   │started│
+ *   └┬──────┘
+ *   ┌▽────────┐
+ *   │  idle   │
+ *   └△───────┬┘
+ *   ┌┴──────┐│
+ *   │running││ (notify_ready())
+ *   └△──────┘│
+ *   ┌┴───────▽─┐
+ *   │ waiting  │
+ *   └──────────┘
+ *
+ * Summary of state transitions
+ * * idle -> waiting  on notify_ready()
+ * * waiting -> running after scheduling
+ * * running -> idle on notify_done()
+ */
+
+class translator_executable {
+public:
+    explicit translator_executable(std::unique_ptr<translator> translator)
+      : _translator(std::move(translator)) {}
+    translator_executable(const translator_executable&) = delete;
+    translator_executable& operator=(const translator_executable&) = delete;
+    translator_executable(translator_executable&& other) noexcept;
+    translator_executable& operator=(translator_executable&&) noexcept;
+    ~translator_executable() = default;
+
+    translation_status status() const;
+    const std::unique_ptr<translator>& translator_ptr() const {
+        return _translator;
+    }
+    const clock::time_point start_time() const { return _start_time; }
+    clock::duration total_wait_time() const;
+    clock::duration total_running_time() const;
+    size_t slices_scheduled() const { return _slices_scheduled; }
+    bool stop_in_progress() const { return _stop_in_progress; }
+
+    void mark_waiting();
+    void mark_running();
+    void mark_idle();
+    void mark_stopping();
+
+    // hook into the waiting queue
+    intrusive_list_hook _waiting_hook;
+    // hook into the running queue
+    intrusive_list_hook _running_hook;
+
+    friend std::ostream&
+    operator<<(std::ostream&, const translator_executable&);
+
+private:
+    friend struct executor;
+    friend class scheduler;
+    std::unique_ptr<translator> _translator;
+    clock::time_point _start_time = clock::now();
+    // Total time since registration the translator has spent in
+    // waiting state.
+    clock::duration _total_waiting_time{0};
+    // Total time since registration the translator has spent in
+    // running state.
+    clock::duration _total_running_time{0};
+    // Total number times a translator was scheduled to
+    // scheduled for translation.
+    size_t _slices_scheduled{0};
```

Comment:
nit: typo in comment. Also, is the "slices" terminology important? Curious if it makes sense to call this "translations scheduled" or "translations started" or something